README.md
# Follow Elasticsearch continuously
[![CircleCI](https://circleci.com/gh/mdreem/elasticsearch_follow.svg?style=svg&circle-token=a53243ea7942ee439f51be3ea4fce2628ed4d58f)](https://circleci.com/gh/mdreem/elasticsearch_follow)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/d192317c5ff74fd7a17dc5c0c2f13317)](https://www.codacy.com/manual/mdreem/elasticsearch_follow?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=mdreem/elasticsearch_follow&amp;utm_campaign=Badge_Grade)
[![Coverage Status](https://coveralls.io/repos/github/mdreem/elasticsearch_follow/badge.svg?branch=master)](https://coveralls.io/github/mdreem/elasticsearch_follow?branch=master)

## Overview

elasticsearch_follow is library helping to query Elasticsearch continuously.

It needs <https://github.com/elastic/elasticsearch-py> as a dependency.

elasticsearch_follow acts as a wrapper for elasticsearch-py and handles various
use-cases, like following logs by polling elasticsearch continuously and fetching
loglines via a generator. It is possible to easily fetch lines surrounding a
given logline.

### How to poll Elasticsearch continuously

The polling logic is implemented in the class ElasticsearchFollow, which needs
an Elasircsearch object from elasticsearch-py. The class Follower takes an
ElasticsearchFollow-object and has a method to create a generator which yields
loglines until all elements of a query have been returned. After this a new
generator has to be created and used.

### How to fetch log-lines from Elasticsearch

To just fetch loglines, one can use ElasticsearchFetch which has
a search_surrounding. This returns a list of lists, where each list contains
the queried loglines and the lines before and after as requested by the parameters 
num_before and num_after.

## Installation

You can install the ``elasticsearch`` package with pip:

```bash
pip install elasticsearch_follow
```

See also: <https://pypi.org/project/elasticsearch-follow/>

## Example usage

This package introduces the command line tool ``es_tail`` which can be used for
following logs written to Elasticsearch and directly fetching log lines by a query.
It is possible to configure the output via an format string.

```bash
# Follow the logs written to the indexes starting with logstash.
# Print the fieds @timestamp and message
es_tail -c "http://localhost:9200" tail --index "logstash*" -f "{@timestamp} {message}" 


# Fetch all logs in the last hour with the field loglevel contains ERROR and fetch the two lines before and after.
# Print the fieds @timestamp and message
es_tail -c "http://localhost:9200" fetch --index "logstash" -f "{@timestamp} {message}" --query loglevel:ERROR -A 2 -B 2 -F "now-1h" 


# It is also possible to print nested fields
es_tail -c "http://localhost:9200" fetch --index "logstash" -f "{@timestamp} {message} {kv[field]} {kv[nested][field]}" -F "now-1h" 
```

The command line options can also be given via environment variables by using the prefix ``ES_TAIL``.
For example
```bash
export ES_TAIL_USERNAME='username'
export ES_TAIL_PASSWORD='password'
es_tail -c http://localhost:9200 tail
```

## Example usage of the library

```python
from elasticsearch import Elasticsearch
from elasticsearch_follow import ElasticsearchFollow, Follower

es = Elasticsearch()
es_follow = ElasticsearchFollow(elasticsearch=es)

# The Follower is used to get a generator which yields new 
# elements until it runs out. time_delta give the number of
# seconds to look into the past.
follower = Follower(elasticsearch_follow=es_follow, index='some-index', time_delta=60)

while True:
    entries = follower.generator()
    for entry in entries:
        print(entry)
    time.sleep(0.1)
```


CHANGELOG.md
# Changelog

## 0.2.4

  * Introduce cli tools.
  * The FormattingProcessor can format output based on fields from Elasticsearch.


elasticsearch_follow/cli.py
#!/usr/bin/env python3

import time
from urllib.parse import urlparse

import certifi
import click
import elasticsearch

from .elasticsearch_fetch import ElasticsearchFetch
from .elasticsearch_follow import ElasticsearchFollow
from .follower import Follower
from .formatting_processor import FormattingProcessor

CONTEXT_SETTINGS = dict(
    help_option_names=['-h', '--help'],
    auto_envvar_prefix='ES_TAIL'
)


class Config:
    def __init__(self):
        self.password = None
        self.username = None
        self.host = None
        self.verbose = False


pass_config = click.make_pass_decorator(Config, ensure=True)


def initialize_es_instance(connect, username, password, verbose=False):
    if verbose:
        print('Attempting to connect to Elasticsearch...')

    http_auth = None
    if username or password:
        http_auth = (username, password)

    if verbose:
        print('Connecting to "{}" with "{}".'.format(connect.netloc, username))

    es = elasticsearch.Elasticsearch(
        [connect.hostname],
        http_auth=http_auth,
        scheme=connect.scheme,
        port=connect.port,
        ca_certs=certifi.where()
    )

    return es


def parse_url(ctx, param, value):
    if value:
        return urlparse(value)


@click.group(context_settings=CONTEXT_SETTINGS)
@click.option('--username', '-u', help='Username for basic-auth.')
@click.option('--password', '-p', help='Password for basic-auth.')
@click.option('--connect', '-c', required=True, callback=parse_url, help='URL to connect to.')
@click.option('--verbose', '-v', is_flag=True, default=False, help='Print more output.')
@pass_config
def cli(config, username, password, connect, verbose):
    config.username = username
    config.password = password
    config.connect = connect
    config.verbose = verbose


@cli.command()
@click.option('--format-string', '-f', default='{@timestamp} {message}', show_default=True, help='Format of the log output.')
@click.option('--index', '-i', help='Determines which indexes to search. Using wildcards is possible.')
@click.option('--num-after', '-A', default=0, type=int, metavar='<NUM>', help='Print <NUM> lines of trailing context after each match.')
@click.option('--num-before', '-B', default=0, type=int, metavar='<NUM>', help='Print <NUM> lines of leading context before each match.')
@click.option('--query', '-q', type=str, help='The query in the Elasticsearch query language.')
@click.option('--from-time', '-F', type=str, help='From which point in time to start the query. Takes an elasticsearch time format. (e.g. now, now-1h).')
@click.option('--to-time', '-T', type=str, help='From which point in time to start the query. Takes an elasticsearch time format. (e.g. now, now-1h). ')
@pass_config
def fetch(config, format_string, index, num_after, num_before, query, from_time, to_time):
    es = initialize_es_instance(config.connect, config.username, config.password, config.verbose)
    es_fetch = ElasticsearchFetch(elasticsearch=es)
    processor = FormattingProcessor(format_string=format_string)

    entries = es_fetch.search_surrounding(index=index, query_string=query, num_before=num_before, num_after=num_after, from_time=from_time, to_time=to_time)
    for entry in entries:
        if num_before > 0 or num_after > 0:
            print('#########')
        for line in entry:
            print(processor.process_line(line))


@cli.command()
@click.option('--format-string', '-f', default='{@timestamp} {message}', show_default=True, help='Format of the log output.')
@click.option('--index', '-i', help='Determines which indexes to search. Using wildcards is possible.')
@click.option('--query', '-q', type=str, help='The query used to filter lines. Has to be given in the Elasticsearch query language.')
@click.option('--number-of-lines', '-n', default=0, type=int, metavar='<NUM>', help='Print <NUM> lines.')
@click.option('--timedelta', '-t', default=60, type=int, metavar='<SECONDS>', help='Look <SECONDS> seconds into the past to update loglines.')
@pass_config
def tail(config, format_string, index, query, number_of_lines, timedelta):
    es = initialize_es_instance(config.connect, config.username, config.password, config.verbose)

    es_follow = ElasticsearchFollow(es, query_string=query)
    follower = Follower(elasticsearch_follow=es_follow, index=index, time_delta=timedelta, processor=FormattingProcessor(format_string=format_string))

    if number_of_lines > 0:
        for entry in list(follower.generator())[-number_of_lines:]:
            print(entry)

    while True:
        entries = follower.generator()
        for entry in entries:
            if entry:
                print(entry)
        time.sleep(0.1)
